### PR TITLE
Eliminating usage of Enum.Parse/Enum.TryParse from critical performance paths

### PR DESCRIPTION
**Pull request by TalZaccai**

### CODE REVIEW

File: libs/cluster/Session/RespClusterFailoverCommands.cs [hunk 1]
```diff
@@ -30,12 +30,15 @@ private bool NetworkClusterFailover(out bool invalidParameters)
...
-                if (!parseState.TryGetEnum(0, ignoreCase: true, out failoverOption) || !failoverOption.IsValid(parseState.GetArgSliceByRef(0).Span))
+                var sbFailoverOption = parseState.GetArgSliceByRef(0).ReadOnlySpan;
+                if (!common.FailoverUtils.TryParseFailoverOption(sbFailoverOption, out failoverOption) ||
+                    failoverOption == FailoverOption.DEFAULT || failoverOption == FailoverOption.INVALID)
                 {
                     var failoverOptionStr = parseState.GetString(0);
 
                     // On failure set the invalid flag, write error and continue parsing to drain rest of parameters if any
-                    while (!RespWriteUtils.WriteError($"ERR Failover option ({failoverOptionStr}) not supported", ref dcurr, dend))
+                    while (!RespWriteUtils.WriteError($"ERR Failover option ({failoverOptionStr}) not supported",
```
Comment (badrishc) [inline libs/cluster/Session/RespClusterFailoverCommands.cs:40]: minor nit: make it a single line as before

**Comment by badrishc** [comment]
Would be nice to have a common pattern for all the enums:

```cs
enum EnumName { ... }
static class EnumNameExtensions
{
   static bool TryParseEnumName(this ArgSlice slice, out ...); // extension method on ArgSlice
}

if !parseState.GetArgSliceByRef(0).TryParseEnumName(...)
   ...
```

**Comment by TalZaccai** [comment]
> Would be nice to have a common pattern for all the enums:
> 
> ```cs
> enum EnumName { ... }
> static class EnumNameExtensions
> {
>    static bool TryParseEnumName(this ArgSlice slice, out ...); // extension method on ArgSlice
> }
> 
> if !parseState.GetArgSliceByRef(0).TryParseEnumName(...)
>    ...
> ```

Maybe have it as an extension method on the parse state? Then it would match up with other similar parse methods (TryGetInt etc).

**Comment by TalZaccai** [comment]
> > > Would be nice to have a common pattern for all the enums:
> > > ```cs
> > > enum EnumName { ... }
> > > static class EnumNameExtensions
> > > {
> > >    static bool TryParseEnumName(this ArgSlice slice, out ...); // extension method on ArgSlice
> > > }
> > > 
> > > if !parseState.GetArgSliceByRef(0).TryParseEnumName(...)
> > >    ...
> > > ```
> > 
> > 
> > Maybe have it as an extension method on the parse state? Then it would match up with other similar parse methods (TryGetInt etc).
> 
> SGTM

Looks cleaner now I think :)

File: libs/server/SessionParseStateExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            else if (sbArg.EqualsUpperCaseSpanIgnoringCase("REPLICATIONMANAGER"u8))
+                value = ManagerType.ReplicationManager;
+            else if (sbArg.EqualsUpperCaseSpanIgnoringCase("SERVERLISTENER"u8))
+                value = ManagerType.ServerListener;
+            else return false;
+
+            return true;
+        }
+    }
+}
```
Comment (badrishc) [inline libs/server/SessionParseStateExtensions.cs:163]: What about the enums we handled manually earlier, such as `TryGetSortedSetAddOption`?

Comment (TalZaccai) [reply libs/server/SessionParseStateExtensions.cs:217]: Added these 2, didn't find any others...

**Review by badrishc — APPROVED** [review]

**Pull request merged**